### PR TITLE
HHH-18371 JDBC connection acquisition failure can lead to query cache corruption

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix https://hibernate.atlassian.net/browse/HHH-18371 .

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/querycache/QueryCacheTest.java [hunk 1]
```diff
@@ -68,7 +76,8 @@
 		settings = {
 				@Setting(name = AvailableSettings.USE_QUERY_CACHE, value = "true"),
 				@Setting(name = AvailableSettings.CACHE_REGION_PREFIX, value = "foo"),
-				@Setting(name = AvailableSettings.USE_SECOND_LEVEL_CACHE, value = "true")
+				@Setting(name = AvailableSettings.USE_SECOND_LEVEL_CACHE, value = "true"),
+				@Setting(name = JdbcSettings.CONNECTION_PROVIDER, value = "org.hibernate.orm.test.querycache.QueryCacheTest$ProxyConnectionProvider")
 		}
 )
 public class QueryCacheTest {
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/orm/test/querycache/QueryCacheTest.java:83]: I tried running this test _without_ your fix, and it did not fail (again current `main` branch).

I think it's important that we have a test which does reliably fail and demonstrate the issue.

**Review by gavinking — CHANGES_REQUESTED** [review]

**Comment by gavinking** [comment]
P.S. Apologies for taking so long to get to this; we have all been on vacation.

File: hibernate-core/src/test/java/org/hibernate/orm/test/querycache/QueryCacheTest.java (hunk 1 above)
Comment (reda-alaoui) [reply hibernate-core/src/test/java/org/hibernate/orm/test/querycache/QueryCacheTest.java:83]: I agree. The test was failing the day I opened the pull request. Maybe something changed. I will check that asap.

Comment (gavinking) [reply hibernate-core/src/test/java/org/hibernate/orm/test/querycache/QueryCacheTest.java:83]: Thanks

**Comment by reda-alaoui** [comment]
@gavinking , I can confirm the issue is fixed in version 6.6.0. I think the test case should be committed nonetheless to prevent any future regression.

**Comment by mbladel** [comment]
Thank you @reda-alaoui, I've rebased and cleaned up your test here https://github.com/hibernate/hibernate-orm/pull/9409

**Comment by reda-alaoui** [comment]
Thanks @mbladel :+1:

**Pull request closed without merging**